torch/fx/utils.py

```diff
@@ -0,0 +1,144 @@
+import torch
+import torch.fx
+import copy
+
+def extract_module(mod : torch.fx.GraphModule, target_qualname : str) -> torch.fx.GraphModule:
+    """
+    Given a GraphModule and a qualified name for a Module resident in the
+    GraphModule's module hierarchy, return a new GraphModule that is equivalent
+    to the original, but with the nodes originating from `target_qualname`
+    out-lined (opposite of inlined).
+    """
+    target_module : torch.nn.Module = mod
+    for atom in target_qualname.split('.'):
+        target_module = getattr(target_module, atom)
+
+
+    # ===== Stage 1: identify Nodes to split off and external data deps =====
+    block_nodes = []
+
+    # Keep track of data dependencies used and referenced in the
+    # extracted block and the rest of the code, respectively.
+    #
+    # We'll use this later to figure out what needs to be inputs/outputs
+    # to/from the extracted block
+    block_defs = set()
+    block_uses = set()
+    rest_defs = set()
+    rest_uses = set()
+
+    for node in mod.graph.nodes:
+        def in_block(node):
+            return node.module_qualname and node.module_qualname.startswith(target_qualname)
+
+        def_set = block_defs if in_block(node) else rest_defs
+        use_set = block_uses if in_block(node) else rest_uses
+        if in_block(node):
+            block_nodes.append(node)
+
+        for arg in node.args:
+            if isinstance(arg, torch.fx.Node):
+                use_set.add(arg)
+        for k, v in node.kwargs.items():
+            if isinstance(v, torch.fx.Node):
+                use_set.add(v)
+        def_set.add(node)
+
+    block_inputs = block_uses - block_defs
+    block_outputs = list(rest_uses - rest_defs)
+    assert len(block_outputs) == 1
```

Comment:
you probably want a more understandable message here. If someone reassigned qualnames (and I bet people will), they can end up in situation when the module is no longer splittable and I suppose they'd hit this assert